### PR TITLE
introduce error messages:

### PR DESCRIPTION
**Pull request by Glignos**
* Show error messages from backend and minor fixes
* Button now is disabled until actions are previewed
Signed-off-by: Glignos <glignos93@gmail.com>

### CODE REVIEW

File: src/app/editor-toolbar-container/editor-toolbar-container.component.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
-import { Component, Output, EventEmitter } from '@angular/core';
+import { Component, Input, Output, EventEmitter } from '@angular/core';
 
 @Component({
   selector: 'me-editor-toolbar-container',
   templateUrl: './editor-toolbar-container.component.html',
   styleUrls: ['./editor-toolbar-container.component.scss']
 })
 export class EditorToolbarContainerComponent {
```
Comment (harunurhan) [inline src/app/editor-toolbar-container/editor-toolbar-container.component.ts]: why this is called container? (I know that this is not part of the PR)

File: src/app/multi-editor/multi-editor.component.ts [hunk 2]
```diff
@@ -75,9 +76,14 @@ export class MultiEditorComponent implements OnInit {
     this.previewedActions = this.userActions;
     this.queryService.previewActions(this.userActions, this.lastSearchedQuery, this.currentPage, this.pageSize)
       .then((res) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts:76]: this endpoint should return an typed object (needs an `interface`), so that we don't have `['property']` access.

File: src/app/multi-editor/multi-editor.component.ts [hunk 3]
```diff
@@ -75,9 +76,14 @@ export class MultiEditorComponent implements OnInit {
...
-        this.newRecords = res['json_records'];
-        this.recordErrors = res['errors'];
-        this.previewMode = true;
+        if (res['response_text']) {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: this is a bad practice, both for backend and frontend.

- If there is an error, return error from backend and handle it in `catch` of promise.
- If you __have to__ return success and body that contains, `response_text` is terrible name for a key that holds the error message.

File: src/app/multi-editor/multi-editor.component.ts [hunk 4]
```diff
@@ -75,9 +76,14 @@ export class MultiEditorComponent implements OnInit {
...
+          this.errorMessage = res['response_text'];
+          this.totalRecords = -1;
+        } else {
+          this.newRecords = res['json_records'];
+          this.recordErrors = res['errors'];
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: what is `recordErrors`?

File: src/app/multi-editor/multi-editor.component.ts [hunk 5]
```diff
@@ -148,9 +154,15 @@ export class MultiEditorComponent implements OnInit {
   private queryCollection(query, collection) {
     this.queryService.searchRecords(query, this.currentPage, collection, this.pageSize)
       .then((json) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/shared/services/query.service.ts [hunk 6]
```diff
@@ -11,10 +11,17 @@ import { UserActions } from '../interfaces';
 export class QueryService {
   readonly url = `${environment.baseUrl}/api/multieditor`;
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: private

File: src/app/shared/services/query.service.ts [hunk 7]
```diff
@@ -11,10 +11,17 @@ import { UserActions } from '../interfaces';
...
   readonly schemaUrl = `${environment.baseUrl}/schemas/records`;
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: private

File: src/app/shared/services/query.service.ts [hunk 8]
```diff
@@ -11,10 +11,17 @@ import { UserActions } from '../interfaces';
...
+  readonly matchTypesMap = {
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: `private`

Comment (harunurhan) [reply src/app/shared/services/query.service.ts]: more important thing is that this is kinda showing a some sort of bad design, I think it's better to unify this both user readable operation names and backend operation names.

File: src/app/shared/services/query.service.ts [hunk 9]
```diff
@@ -24,6 +31,7 @@ export class QueryService {
   }
 
   previewActions(userActions: UserActions, queryString: string, pageNum: number, pageSize: number): Promise<object[]> {
+    this.resolveMatchMap(userActions);
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: if you have to use this, this should return. input-output function parameters are bad and we should avoid them.

File: src/app/actions/actions.component.ts [hunk 10]
```diff
@@ -19,20 +19,24 @@ export class ActionsComponent implements OnInit {
...
+    return this.appGlobalsService.actionMatchTypes;
+  }
+
+  get conditionMatchTypes(): string[] {
+    return this.appGlobalsService.conditionMatchTypes;
   }
 
   addAction() {
-    this.userActionsService.addDefaultAction(this.selectedAction, this.matchTypes[0]);
+    this.userActionsService.addDefaultAction(this.selectedAction, this.actionMatchTypes[0]);
```
Comment (harunurhan) [inline src/app/actions/actions.component.ts]: why you are passing a selected action and default match type to a method which is called `addDefaultAction` this picking the default should be inside of this function.

File: src/app/actions/actions.component.ts [hunk 11]
```diff
@@ -19,20 +19,24 @@ export class ActionsComponent implements OnInit {
...
   }
 
   addCondition() {
-    this.userActionsService.addDefaultCondition(this.matchTypes[0]);
+    this.userActionsService.addDefaultCondition(this.conditionMatchTypes[0]);
```
Comment (harunurhan) [inline src/app/actions/actions.component.ts]: ditto

File: src/app/actions/actions.component.ts [hunk 12]
```diff
@@ -19,20 +19,24 @@ export class ActionsComponent implements OnInit {
 
   ngOnInit() {
     this.userActions = this.userActionsService.getUserActions();
-    this.userActionsService.addDefaultAction(this.selectedAction, this.matchTypes[0]);
-    this.userActionsService.addDefaultCondition(this.matchTypes[0]);
+    this.userActionsService.addDefaultAction(this.selectedAction, this.actionMatchTypes[0]);
```
Comment (harunurhan) [inline src/app/actions/actions.component.ts]: ditto

File: src/app/actions/actions.component.ts [hunk 13]
```diff
@@ -19,20 +19,24 @@ export class ActionsComponent implements OnInit {
...
+    this.userActionsService.addDefaultCondition(this.conditionMatchTypes[0]);
```
Comment (harunurhan) [inline src/app/actions/actions.component.ts]: ditto

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts [hunk 14]
```diff
@@ -1,8 +1,13 @@
...
+import { Component, Input, Output, EventEmitter } from '@angular/core';
+
 
 @Component({
   selector: 'me-editor-toolbar-save',
   templateUrl: './editor-toolbar-save.component.html',
   styleUrls: ['../../multi-editor/multi-editor.component.scss']
 })
-export class EditorToolbarSaveComponent { }
+export class EditorToolbarSaveComponent {
```
Comment (harunurhan) [inline src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts]: I am not sure if this supposed to be a component.

**Review by harunurhan — CHANGES_REQUESTED** [review]
General comment: drop `on` prefix for `@Output`s.
#20

File: src/app/editor-toolbar-container/editor-toolbar-container.component.ts (hunk 1 above)
Comment (Glignos) [reply src/app/editor-toolbar-container/editor-toolbar-container.component.ts]: @zzacharo can you check if this should change?

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts (hunk 14 above)
Comment (Glignos) [reply src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts]: @zzacharo  ditto

File: src/app/editor-toolbar-container/editor-toolbar-container.component.ts (hunk 1 above)
Comment (zzacharo) [reply src/app/editor-toolbar-container/editor-toolbar-container.component.ts]: Because it contains `Save` and `Search` component.

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts (hunk 14 above)
Comment (zzacharo) [reply src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts]: This is gonna have some save logic in the future.

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts [hunk 15]
```diff
@@ -1,8 +1,13 @@
...
+  @Input() canSave: boolean;
+  @Output() onSubmitActions = new EventEmitter<void>();
```
Comment (zzacharo) [inline src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts]: Couldn't you call submit here? @Glignos

File: src/app/multi-editor/multi-editor.component.ts [hunk 16]
```diff
@@ -74,14 +75,15 @@ export class MultiEditorComponent implements OnInit {
   onPreviewClick() {
     this.previewedActions = this.userActions;
     this.queryService.previewActions(this.userActions, this.lastSearchedQuery, this.currentPage, this.pageSize)
-      .then((res) => {
-        this.newRecords = res['json_records'];
-        this.recordErrors = res['errors'];
+      .then((res: RecordsPreview) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: you don't have to put the type here, if `previewActions` has it already.

File: src/app/multi-editor/multi-editor.component.ts [hunk 17]
```diff
@@ -135,26 +137,29 @@ export class MultiEditorComponent implements OnInit {
   private getNewPageRecords() {
     this.queryService
       .fetchNewPageRecords(this.previewedActions, this.lastSearchedQuery, this.currentPage, this.lastSearchedCollection, this.pageSize)
-      .subscribe((json) => {
-        this.records = json.oldRecords['json_records'];
-        this.uuids = json.oldRecords['uuids'];
-        this.newRecords = json.newRecords['json_records'];
-        this.recordErrors = json.newRecords['errors'];
+      .subscribe((json: RecordsToCompare) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/multi-editor/multi-editor.component.ts [hunk 18]
```diff
@@ -135,26 +137,29 @@ export class MultiEditorComponent implements OnInit {
...
+        this.validationErrors = json.newRecords.errors;
         this.changeDetectorRef.markForCheck();
       },
       error => { this.errorText = error; this.changeDetectorRef.markForCheck(); });
   }
 
   private queryCollection(query, collection) {
     this.queryService.searchRecords(query, this.currentPage, collection, this.pageSize)
-      .then((json) => {
+      .then((json: QueryResult) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: ditto

File: src/app/shared/interfaces/query-result.ts [hunk 19]
```diff
@@ -0,0 +1,5 @@
+export interface QueryResult {
+    json_records: object[];
+    uuids: string[];
+    total_records: number;
+};
```
Comment (harunurhan) [inline src/app/shared/interfaces/query-result.ts]: don't you have linter how's that passing, unnecessary `;`

File: src/app/shared/interfaces/records-preview.ts [hunk 20]
```diff
@@ -0,0 +1,4 @@
+export interface RecordsPreview {
+    json_records: object[];
+    errors: string[];
+  };
```
Comment (harunurhan) [inline src/app/shared/interfaces/records-preview.ts]: ditto

File: src/app/shared/interfaces/records-to-compare.ts [hunk 21]
```diff
@@ -0,0 +1,5 @@
+import { QueryResult, RecordsPreview} from '../interfaces';
+export interface RecordsToCompare {
+    oldRecords: QueryResult;
+    newRecords: RecordsPreview;
+};
```
Comment (harunurhan) [inline src/app/shared/interfaces/records-to-compare.ts]: ditto

File: src/app/shared/interfaces/records-to-compare.ts [hunk 22]
```diff
@@ -0,0 +1,5 @@
+import { QueryResult, RecordsPreview} from '../interfaces';
```
Comment (harunurhan) [inline src/app/shared/interfaces/records-to-compare.ts]: space

File: src/app/multi-editor/multi-editor.component.ts [hunk 23]
```diff
@@ -135,26 +137,29 @@ export class MultiEditorComponent implements OnInit {
...
         this.totalRecords = json['total_records'];
         this.uuids = json['uuids'];
         this.changeDetectorRef.markForCheck();
       })
       .catch(error => {
-        this.errorText = error;
+        if (error.json().message) {
+          this.customError = error.json().message;
+        }
         this.changeDetectorRef.markForCheck();
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: inside the `if`

File: src/app/multi-editor/multi-editor.component.ts [hunk 24]
```diff
@@ -103,7 +105,7 @@ export class MultiEditorComponent implements OnInit {
   onCollectionChange(selectedCollection: string) {
     this.selectedCollection = selectedCollection;
     this.queryService.fetchCollectionSchema(this.selectedCollection)
-      .then((res) => {
+      .then((res: object) => {
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `res: object` => `schema`

File: src/app/multi-editor/multi-editor.component.ts [hunk 25]
```diff
@@ -6,7 +6,7 @@ import {
 import 'rxjs/add/operator/toPromise';
 
 import { SchemaKeysStoreService, QueryService } from '../shared/services';
-import { UserActions } from '../shared/interfaces';
+import { UserActions, QueryResult, RecordsPreview, RecordsToCompare } from '../shared/interfaces';
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: don't forget the clean imports once you removed the types here.

File: src/app/multi-editor/multi-editor.component.ts [hunk 26]
```diff
@@ -22,10 +22,11 @@ export class MultiEditorComponent implements OnInit {
   totalRecords = -1;
   schema: object;
   errorText: string;
-  recordErrors: string[];
+  validationErrors: string[];
   lastSearchedQuery = '';
   lastSearchedCollection: string;
   previewedActions: UserActions;
+  customError: string;
```
Comment (harunurhan) [inline src/app/multi-editor/multi-editor.component.ts]: `errorMessage` or even `httpErrorMessage`

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts [hunk 27]
```diff
@@ -1,8 +1,13 @@
...
 
 @Component({
   selector: 'me-editor-toolbar-save',
   templateUrl: './editor-toolbar-save.component.html',
   styleUrls: ['../../multi-editor/multi-editor.component.scss']
 })
-export class EditorToolbarSaveComponent { }
+export class EditorToolbarSaveComponent {
+  @Input() canSave: boolean;
+  @Output() submitActions = new EventEmitter<void>();
```
Comment (harunurhan) [inline src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.ts]: `submit` or `save` or `saveClick`

File: src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.html [hunk 28]
```diff
@@ -1 +1 @@
-<button type="button" class="btn btn-default navbar-btn btn-success"><i class="fa fa-floppy-o" aria-hidden="true"></i>Save</button>
+<button type="button" [disabled]="!canSave" (click)="onSubmitActions.emit()" class="btn btn-default navbar-btn btn-success"><i class="fa fa-floppy-o" aria-hidden="true"></i>Save</button>
```
Comment (harunurhan) [inline src/app/editor-toolbar-container/editor-toolbar-save/editor-toolbar-save.component.html]: `onSumbitActions` doesn't exist

File: src/app/editor-toolbar-container/editor-toolbar-container.component.ts [hunk 29]
```diff
@@ -1,13 +1,15 @@
...
+  @Input() canSave: boolean;
   @Output() onSearchRecordClick = new EventEmitter<string>();
   @Output() onCollectionChange = new EventEmitter<string>();
+  @Output() submitActions = new EventEmitter<void>();
```
Comment (harunurhan) [inline src/app/editor-toolbar-container/editor-toolbar-container.component.ts]: same renaming

File: src/app/editor-toolbar-container/editor-toolbar-container.component.ts (hunk 1 above)
Comment (harunurhan) [reply src/app/editor-toolbar-container/editor-toolbar-container.component.ts]: `Toolbar` would be more simple.

**Review by harunurhan — CHANGES_REQUESTED** [review]

**Comment by Glignos** [comment]
@harunurhan check again?

File: src/app/shared/interfaces/records-to-compare.ts [hunk 30]
```diff
@@ -0,0 +1,5 @@
+import { QueryResult, RecordsPreview } from '../interfaces';
+export interface RecordsToCompare {
```
Comment (harunurhan) [inline src/app/shared/interfaces/records-to-compare.ts:2]: what is this?

File: src/app/shared/services/query.service.ts [hunk 31]
```diff
@@ -34,7 +34,8 @@ export class QueryService {
       .toPromise();
   }
 
-  fetchNewPageRecords(userActions: UserActions, queryString: string, page: number, collection: string, pageSize: number): Observable<any> {
+  fetchNewPageRecords(userActions: UserActions, queryString: string, page: number,
```
Comment (harunurhan) [inline src/app/shared/services/query.service.ts]: have the param consistency with the `searchRecords` same order same names.

**Review by harunurhan — CHANGES_REQUESTED** [review]
dont push package-lock.json for now, we still haven't decided to use it, add it to gitignore if you want.

File: src/app/shared/interfaces/records-to-compare.ts (hunk 30 above)
Comment (Glignos) [reply src/app/shared/interfaces/records-to-compare.ts:2]: Its an interface for the Preview function that fetches the old and new records to see the changes.

**Review by harunurhan — APPROVED** [review]
